File: 프로그래머스/unrated/178871. 달리기 경주/달리기 경주.java
import java.util.Map;
import java.util.HashMap;
import java.util.TreeMap;
class Solution {
    public String[] solution(String[] players, String[] callings) {
        int N = players.length;
        String[] answer = new String[N];
        Map<Integer, String> treemap = new TreeMap<>();//정렬된 값
        Map<String, Integer> hashmap = new HashMap<>();//idx 정보 찾기
        for(int i = 0; i < N; i++){
            treemap.put(i,players[i]);
            hashmap.put(players[i],i);
        }
        for(String calling : callings){
            int idx = hashmap.get(calling);//앞서 나갈 사람의 이름이 담긴 idx(idx-1)로 나아감
            String backplayer = treemap.get(idx-1);//뒤쳐진 사람과 앞서 나간 사람 Swap;
            
            //hashmapSwap;
            hashmap.put(calling,idx-1);//앞서 가는 자.
            hashmap.put(backplayer,idx);//뒤로 간 자.
            //treemapSwap;
            treemap.put(idx-1,calling);//앞서는 사람 순으로 TreeMap에서 자동 정렬
            treemap.put(idx,backplayer);//
        }
        int j = 0;
        for(String value : treemap.values()){
            answer[j++] = value;
        }
        return answer;
    }
}

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

typedef struct _node {
    int idx;
    struct _node* nx[26];
} node;
node* create_node() {
    node* np = (node*)malloc(sizeof(node));
    for(int i = 0; i < 26; i++) np->nx[i] = NULL;
    return np;
}
void delete_node(node* np) {
    for(int i = 0; i < 26; i++) {
        if(np->nx[i]) delete_node(np->nx[i]);
    }
    free(np);
}
void add_node(node* root, char* str, int idx) {
    node* np = root;
    while(*str) {
        int i = *str++ - 'a';
        if(!np->nx[i]) np->nx[i] = create_node();
        np = np->nx[i];
    }
    np->idx = idx;
}
int find_idx(node* root, char* str) {
    node* np = root;
    while(*str) {
        int i = *str++ - 'a';
        np = np->nx[i];
    }
    return np->idx;
}
void swap(char** a, char** b) {
    char* tmp = *a;
    *a = *b;
    *b = tmp;
}
char** solution(const char* players[], size_t players_len, const char* callings[], size_t callings_len) {
    char** answer = (char**)malloc(players_len * sizeof(char*));
    node* root = create_node();
    for(int i = 0; i < players_len; i++) {
        answer[i] = (char*)malloc((strlen(players[i]) + 1) * sizeof(char));
        strcpy(answer[i], players[i]);
        add_node(root, players[i], i);
    }
    for(int i = 0; i < callings_len; i++) {
        int idx = find_idx(root, callings[i]);
        add_node(root, answer[idx], idx - 1);
        add_node(root, answer[idx - 1], idx);
        swap(&answer[idx], &answer[idx - 1]);
    }

    delete_node(root);
    return answer;
}